blockscout-ens/bens-logic/src/protocols/protocoler.rs

```diff
@@ -289,15 +290,60 @@ impl Protocoler {
         network_id: i64,
         maybe_filter: Option<NonEmpty<String>>,
     ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
-        let tld = Tld::from_domain_name(name).ok_or_else(|| ProtocolError::InvalidName {
-            name: name.to_string(),
-            reason: "no tld found".to_string(),
-        })?;
+        let tlds = self
+            .networks
+            .get(&network_id)
+            .ok_or_else(|| ProtocolError::NetworkNotFound(network_id))?
+            .use_protocols
+            .iter()
+            .filter_map(|protocol_name| {
+                self.protocols
+                    .get(protocol_name)
+                    .map(|protocol| protocol.info.tld_list.iter().cloned())
+            })
+            .flatten()
+            .collect::<Vec<Tld>>();
+
+        if name.contains('.') {
+            return self.find_names_with_tld(name, network_id, maybe_filter);
+        }
+
+        let mut all_names_with_protocols = Vec::new();
+        for tld in tlds {
+            if all_names_with_protocols.len() >= MAX_NAMES_LIMIT {
+                break;
+            }
+            let name_with_tld = format!("{}.{}", name, tld.0);
+            if let Ok(mut names) =
+                self.find_names_with_tld(&name_with_tld, network_id, maybe_filter.clone())
+            {
+                all_names_with_protocols.append(&mut names);
+            }
+        }
+
+        if all_names_with_protocols.is_empty() {
+            return Err(ProtocolError::InvalidName(name.to_string()));
+        }
+
+        Ok(all_names_with_protocols
+            .into_iter()
+            .take(MAX_NAMES_LIMIT)
+            .collect())
+    }
+
+    pub fn find_names_with_tld(
```

Comment:
why `find_names_with_tld` ? you don't search anything here, you just return name options

Reply:
name_options_for_tld

Reply:
I think it's better to leave old function name here and rename new function to `names_options_in_network_with_suggestions` or smthg like that